lib/modules.nix

```
@@ -52,6 +53,286 @@ in
+            options = {
+              enable = lib.mkOption {
+                type = lib.types.bool;
+                default = enableProvider;
+                example = !enableProvider;
+                description = ''
+                  Whether to integrate this plugin with blink.cmp.
+                '';
+              };
+              key = lib.mkOption {
+                type = lib.types.str;
+                default = key;
+                description = ''
+                  The key to use for ${pluginName}'s blink.cmp provider.
+                  This is the id you should use when including this provider in completion source lists.
+                  Must be unique.
+                '';
+              };
+              default = lib.mkOption {
+                type = lib.types.bool;
+                default = enableDefault;
+                example = !enableDefault;
+                description = ''
+                  Whether to include this plugin in the `default` completion source list.
+                '';
+              };
+              cmdline = lib.mkOption {
+                type = lib.types.bool;
+                default = enableCmdline;
+                example = !enableCmdline;
+                description = ''
+                  Whether to include this plugin in the `cmdline` completion source list.
+                '';
+              };
+              filetypes = lib.mkOption {
+                type = lib.types.attrsOf lib.types.bool;
+                # Only include `true` attrs in the final value
+                apply = lib.filterAttrs (_: lib.id);
+                default = enabledFiletypes;
+                # TODO: example
+                description = ''
+                  Whether to include this plugin in the specific `per_filetype` completion source lists.
+                '';
+              };
+              settings = lib.mkOption {
+                default = { };
+                description = ''
+                  Settings for the blink.cmp completion provider.
+                '';
+                example = settingsExample;
+                type = lib.types.submodule [
+                  {
+                    options = {
+                      enabled = internal;
+                      module = internal;
+                    };
+                    config = {
+                      name = lib.mkDefault sourceName;
+                      inherit module;
+                      # TODO: use mkDerivedConfig to preserve definition override priority?
+                      opts = lib.mkIf usePluginSettings pluginCfg.settings;
+                    };
+                  }
+                  (import ../plugins/by-name/blink-cmp/provider-config.nix lib)
+                ];
+              };
+            };
+          };
+        };
+      };
+      config = lib.mkIf (pluginCfg.enable && cfg.enable) {
+        plugins.blink-cmp.settings.sources = {
+          providers.${cfg.key} = cfg.settings;
```

```suggestion
          providers.${cfg.key} = lib.mkDefault cfg.settings;
```

Probably want to make them default to not break existing configs and support overriding from blink module?

---

> Just checking branch out to test stuff out: 

Anything you're playing with that we could add as test cases?

> ```bash
> … while evaluating the option `plugins.blink-cmp.settings.sources.providers.copilot.async':
> 
>        (stack trace truncated; use '--show-trace' to show the full, detailed trace)
> 
>        error: The option `plugins.blink-cmp.settings.sources.providers.copilot.async` is defined both null and not null, in `/nix/store/k3vjx39q6ksdh9m1yq9n9ikscd5d6387-source/plugins/by-name/blink-copilot' and `/nix/store/v309l0bfp6jis58v9v6a6b8rica9c72z-source/modules/nixvim/plugins/blink'.
> ```

Ah, that'll be because I'm copying `settings` around without preserving the original definition priorities.

Should be fixed by using [`mkAliasDefinitions`](https://github.com/NixOS/nixpkgs/blob/ba2d98239e995de78369da2a9e280ecc3c402e50/lib/modules.nix#L1076-L1097)

---

Just using my neovim flake, as-is. Which enables the plugins but configures them through blink.